Use bitvectors instead of ints where appropriate in the symbolic pipeline

KinanBab opened the pull request:
* Define TypedExpr, our own wrapper class around z3::expr, that maintains sort integrity invariant.
* Overload operated in TypedExpr similar to z3::expr API.
* Use TypedExpr in our symbolic pipeline instead of z3::expr.
* Metadata fields are detected automatically from bvm2 json, with their bitwidth, and created as TypedExpr.
* Further assignment to metadata fields must satisify the invariant.
* Concrete fields are now strings extracted using z3::expr::to_string()
* Support hex string literal values in action bodies.
* Support default table entries.
* Fix: no error thrown when a table has no entries.

KinanBab committed:
First stab at making all things bit vector.

* Headers, metadata, and trace variables are now bitvectors and std::string in their symbolic and concrete counterparts.
* Free headers, metadata, and trace variables are created as bv_const of size '9'
* IrValue are parsed into bitvector using bv_val, also of size '9'
* When evaluating the model, outputs are extracted via get_decimal_string()

TODO:
* Do not hard code 9, use proper sizing for every field and variable.
* Table value should also become a bitvector
* get_decimal_string() should be passed the proper precision (0?) and turned to hex
* IrValues should be parsed into bitvector of an appropriate size, will require padding operations.

KinanBab committed:
Remove hardcoded bit sizes

KinanBab committed:
Add TypedExpr class to wrap z3::expr with fixed sort and sign information. Implement basic functionality for it.

KinanBab committed:
Replace uses of z3::expr with TypeExpr in symbolic pipeline.

TODO:
1. Test util::IrValueToZ3Expr with negative values from p4info and table entries.
2. Improve metadata by looking at bmv2 json header definitions and creating their variables at the start with type information.
3. Test assignment of expressions with wrong sorts to maps (scope, headers, metadata etc).
4. Find a better way to encode undefined initial values.

KinanBab committed:
Support tables without entries

KinanBab committed:
Improve handling of hex string values and passing them to z3

KinanBab committed:
Support tables default entries

KinanBab committed:
Support hex string literal expressions

KinanBab committed:
Update golden test files

KinanBab committed:
Update IR definition, parsing, and golden test files so that we capture both a Header instance definition and its header type definition.

KinanBab committed:
Symbolic pipeline does not hardcode metadata fields and sizes, and instead discovers them by looking at metadata instance definitions and their types

KinanBab committed:
Code formatting

KinanBab committed:
Add validity bit to header/metadata structs

KinanBab committed:
Remove sign completely, simplify and better document TypedExpr

KinanBab commented on p4_symbolic/symbolic/headers.cc:
```diff
@@ -0,0 +1,126 @@
+namespace p4_symbolic {
+namespace symbolic {
+namespace headers {
+
```
This is going to change: we will not have a FreeSymbolicHeader function anymore. SymbolicHeader will not be passed around internally. We will only use SymbolicMetadata, and at the end of the evaluation, we will extract SymbolicHeader from SymbolicMetadata to make it easier for client code to refer to these packet fields in assertions. Same with ConcereteHeader.

konne88 replied:
It's ok with me if we remove all of these, and replace them with hash maps

KinanBab replied:
already done on the other branch :D

KinanBab commented on p4_symbolic/symbolic/headers.h:
```diff
@@ -0,0 +1,46 @@
+// Copyright 2020 Google LLC
```
Ignore this file,  our treatment of headers is going to change.

KinanBab commented on p4_symbolic/symbolic/util.cc:
```diff
@@ -37,189 +43,149 @@ bool Z3BooltoBool(Z3_lbool z3_bool) {
   SymbolicMetadata metadata;
-  metadata.insert({"standard_metadata.ingress_port", ingress_port});
-  metadata.insert({"standard_metadata.egress_spec", Z3Context().int_val(-1)});
+  for (const auto &[header_name, header_type] : headers) {
```
This is where we look up the header type definitions from the bmv2 json, and create free variables to every field defined there, with the same bitwidth.

KinanBab replied:
The pipeline down stream is going to be copying metadata around, and assigning new expressions to the fields in it as the p4 program does. Trigged the copy assignment overloaded operation.

KinanBab committed:
Make the struct with the hardcoded packet field names a helper for external usage.

* Concrete and symbolic structs with these hardocded field names renamed to ConcretePacket and SymbolicPacket (used to be called Headers).
* Generic Map from fully qualified field names to their values renamed to (Symbolic|Concrete)Headers to agree with the P4 and bmv2 JSON terminology.
* Internal SymbolicPerPacketState (used to contain both Headers and Packet helpers) is removed, and replaced with direct usage of SymbolicHeaders, the packet fields were unused because they do not appear *literally* in the p4 program.
* Change Symbolic pipeline code to receive and return SymbolicHeader instead of the old useless wrapper SymbolicPerPacketState type.
* Remove headers.cc and headers.h, replace them with packet.cc and packet.h responsible for extracting the helper structs from the generic headers map (symbolic and concrete).
* Simplify util.cc and util.h, they no longer need to worry about merging packet helper structs or manipulating them, instead they only deal with (Symbolic|Concrete)Headers.
* TODO: Specify the actual mapping between helper packet field names and their corresponding header fields in the p4 program.

KinanBab committed:
Modify symbolic evaluation flow to match flow of p4 program

* Add EvaluateControl helper for evaluating control constructs (tables, and conditionals) and returning their traces.
* Evaluate initial_control first, symbolic.cc no longer goes over all table and evaluates them.
* EvaluateTable works as before, but at the end, it calls EvaluateControl on the next control to evaluate the next flow.
* EvaluateTable adds its table match to the returned trace of the next flow, and returns the result.
* Implement EvaluateConditional which works similar to EvaluateTable, but it calls EvaluateControl twice (once on every branch) and merged
resulting trace using a symbolic if-then-else.
* Finally, add the remaining sample programs as symbolic tests, these do not work fully yet cause they need some more feature support (e.g. ==)

KinanBab committed:
Add support for more RValues and RExpressions

* Support for ==, !=, &&, ||
* Suppport for !
* Support for ternary expressions
* Support for boolean and string literals
* Fix initial expressions in a symbolic table match

KinanBab committed:
Improve main.cc to find packets hitting default entries of tables. Add reflector and hardcoded p4 samples as symbolic tests

konne88 commented on p4_symbolic/symbolic/typed.cc:
```diff
@@ -0,0 +1,97 @@
+
+// Pad if needed by using the appropriate z3 API.
+#define PAD(bitvector, pad_size)                                               \
+  pad_size > 0 ? z3::concat(Z3Context().bv_val(0, pad_size), bitvector.expr()) \
```
Don't define macros (see style guide). Just make this a function.

konne88 commented on p4_symbolic/symbolic/typed.cc:
```diff
@@ -0,0 +1,97 @@
+  pad_size > 0 ? z3::concat(Z3Context().bv_val(0, pad_size), bitvector.expr()) \
+               : bitvector.expr()
+
+// This macro is used through out the implementation of the overloaded
```
See above

konne88 commented on p4_symbolic/symbolic/typed.h:
```diff
@@ -0,0 +1,134 @@
+// bitvector), as well as size information (e.g. the size of the bitvector).
+//
+// Example:
+// // A metadata map from field names to expressions is defined at the begining.
```
This double commenting is weird. maybe indent with 4 spaces to make it less weird

KinanBab replied:
cool

konne88 commented on p4_symbolic/symbolic/typed.h:
```diff
@@ -0,0 +1,134 @@
+  TypedExpr &operator=(TypedExpr &&other);
+
+  // Accessors
+  const z3::expr &expr() const { return this->expr_; }
```
Follow style guide for naming.

konne88 commented on p4_symbolic/symbolic/typed.h:
```diff
@@ -0,0 +1,134 @@
+  TypedExpr operator!() const;
+
+  // If-then-else.
+  static TypedExpr ite(const TypedExpr &condition, const TypedExpr &true_value,
```
Ite

konne88 commented on p4_symbolic/symbolic/typed.cc:
```diff
@@ -0,0 +1,97 @@
+  if (this->sort().is_bv()) {
+    // We can only re-assign bit vectors to ones of equal or larger size.
+    unsigned int size_diff = this->sort().bv_size() - other.sort().bv_size();
+    assert(size_diff >= 0);
```
This is not great. It will just crash the program. It's also not how we handle errors in other cases.

konne88 commented on p4_symbolic/symbolic/typed.h:
```diff
@@ -0,0 +1,134 @@
+// Copyright 2020 Google LLC
```
Looking at this, I would lean towards not introducing this (though I'm not 100% sure it's the right call).

Advantages:
- Assignments are safer.

Disadvantages:
- It provides a false sense of security. Code can still change the bit-width of a metadata field.
- We have to use assertions to fail, which doesn't match what we do anywhere else
- It violates the style guide for &&, || etc
- It adds complexity (it needs more code).
- TypedExpr is a bad name, because z3::Expr is already typed. I can't think of a good name.

I would recommend to instead define the following function over z3::Expr which return a StatusOr<z3::Expr>:

Assign
And
Ife
...

One question. How often do we actually assign? If it's 10 times, the case for introducing this is much weaker than, if say, it was 100 times.

KinanBab committed:
Support remaining binary and unary operators

KinanBab committed:
Use match target expression from bmv2 when evaluating a match instead of the match string name from p4info

KinanBab committed:
Support builtin expressions for data type conversions

KinanBab committed:
Implement LPM match and add ipv4-routing p4-sample as a test for the symbolic pipeline

* Update IR test golden files to contain table match target expression added to table implementation.
* Update output and smt symbolic test golden files to reflect recent changes to symbolic pipeline

KinanBab committed:
Contexts store both ingress and egress headers. main.cc outputs packets in a consistent deterministic order.

KinanBab committed:
Simplify Headers construction and mutation.

* Headers are now an instance of a new class: GuardedMap
* Headers are passed by pointer to functions that may change them
* No functions return a Headers instance anymore, this avoid needless copying
* Sort checked and padding-aware operations are moved to operators.h|cc
* These operations are now proper functions and do not use operator overloading
* These operations return StatusOr to signify sort checking failures, instead of a runtime crash
* Finally, to avoid messing up the headers, since it is now a single instance passed by pointer, we construct and pass guards that are uses to condition assignments to the headers, so they only take effect if the guard is satisified.
* Guards are constructed in lieu of calling util::MergeHeadersOnCondition(), which is now removed
* Updated the SMT program golden tests to reflect the new smt programs

* TODO: use the same guard that is constructed and passed for trace, maybe combine trace and headers as a global object passed around by reference
* TODO: potentially do the above only for table_matches, dropped is an alias and should not be part of the pointed struct

konne88 commented on p4_symbolic/symbolic/action.cc:
```diff
@@ -17,20 +17,26 @@
+    }
+    case ir::Statement::kDrop: {
+      RETURN_IF_ERROR(
+          headers->Set("$dropped$", Z3Context().bool_val(true), guard));
```
Pull this out into a constant?

KinanBab replied:
Good idea

konne88 commented on p4_symbolic/symbolic/action.cc:
```diff
@@ -89,14 +85,26 @@ gutil::StatusOr<SymbolicPerPacketState> EvaluateAssignmentStatement(
 gutil::StatusOr<z3::expr> EvaluateRValue(const ir::RValue &rvalue,
-                                         const SymbolicPerPacketState &state,
+                                         const SymbolicHeaders &headers,
                                          ActionContext *context) {
```
Can this be const?

KinanBab replied:
Yeah it can, let me fix that

konne88 commented on p4_symbolic/symbolic/guarded_map.h:
```diff
@@ -0,0 +1,99 @@
+namespace p4_symbolic {
+namespace symbolic {
+
+// The GuardedMap class wraps around an internal std::unordered_map instance,
```
Great comment!

konne88 commented on p4_symbolic/symbolic/guarded_map.h:
```diff
@@ -0,0 +1,99 @@
+// 2. A value mapped by a key allows has the same sort.
+// 3. A value can only be assigned to a key given a guard.
+//
+class GuardedMap {
```
GuardedMap is a bit too generic, I think. I would just call it SymbolicPerPacketState, but maybe it I think it should have at least Symbolic in the name.

KinanBab replied:
Ah that makes sense, because then it would correspond to the old type name. Now I see why you asked me that on the chat!

KinanBab replied:
Ok so I have given this some thought. I think GuardedMap as a class is a bit more abstract than the concept of SymbolicPerPacketState and SymbolicHeaders.

So I am thinking I can renamed the GuardedMap to SymbolicGuardedMap

And then have two types (that are aliases to SymbolicGuardedMap for now, but may change in the future if more features are added) defined in symbolic.h:
* SymbolicHeaders used in the trace and in other external API related to assertions.
* SymbolicPerPacketState used internally to pass the "state/context" around between evaluation functions.

KinanBab replied:
I renamed things as above. Please let me know if you think it is a good way to organize this.

KinanBab replied:
I went back and removed SymbolicHeaders. Everything uses SymbolicPerPacketState. I renamed ConcreteHeaders to ConcretePerPacketState for symmetry.

konne88 commented on p4_symbolic/symbolic/guarded_map.h:
```diff
@@ -0,0 +1,99 @@
+// 3. A value can only be assigned to a key given a guard.
+//
+class GuardedMap {
+ private:
```
This goes to the end of the class.

konne88 commented on p4_symbolic/symbolic/guarded_map.h:
```diff
@@ -0,0 +1,99 @@
+ private:
+  // The underlying map storing the keys and their values.
+  std::unordered_map<std::string, z3::expr> map_;
+  absl::Status status_;
```
I have no clue why you use this, but my prediction is that it's a bad idea :)

konne88 commented on p4_symbolic/symbolic/guarded_map.cc:
```diff
@@ -0,0 +1,82 @@
+namespace p4_symbolic {
+namespace symbolic {
+
+GuardedMap::GuardedMap(
```
Since your constructor can't fail, use a factory.

StatusOr<GuardedMap> CreateGuardedMap(const google::protobuf::Map<std::string, ir::HeaderType> &headers)

and your (private) constructor then just takes whatever type this->map_ has (so it can't fail).

KinanBab replied:
Neat! I read that in the style guide this morning, that's pretty neat.

konne88 commented on p4_symbolic/symbolic/guarded_map.h:
```diff
@@ -0,0 +1,99 @@
+  GuardedMap &operator=(GuardedMap &&other) = delete;
+
+  // Getters.
+  size_t Count(const std::string &key) const;
```
Super confusing that this returns a size. Just call it ContainsKey

konne88 replied:
Only keep this if it's used by users of the class. Remove if it's only used inside the class (then just inline).

KinanBab replied:
It is used outside the class, I will make it a bool, I was trying to make this look as close as possible to std::unordered_map class.

konne88 commented on p4_symbolic/symbolic/action.h:
```diff
@@ -56,33 +55,52 @@ struct ActionContext {
-    ActionContext *context);
+gutil::StatusOr<z3::expr> EvaluateFieldValue(const ir::FieldValue &field_value,
+                                             const SymbolicHeaders &headers,
+                                             ActionContext *context);
```
why is the context mutable? same below

KinanBab replied:
I will change this to const throughout the various expressions. It only needs to be a pointer for statements.

konne88 commented on p4_symbolic/symbolic/symbolic.h:
```diff
@@ -37,91 +38,37 @@ namespace symbolic {
   // if matched is false, these two fields are set to -1.
   int entry_index;
-  int value;
+  std::string value;
```
Does value even make sense anymore in your world of actions?

konne88 replied:
If not, we can probably just delete it.

KinanBab replied:
Honestly, I am unsure. I have left a couple of TODO comments about this, and I am planning on taking a closer look at this next. I kept it for compatibility with the old Hercules API that we discussed some time ago. I think it can probably go away. But I wanted to understand this a bit better before making the change.

konne88 replied:
I think you should get rid of it. It doesn't make sense anymore.

KinanBab replied:
done.

KinanBab committed:
Clean up TableEntry evaluation. Clean up trace construction and trace merging.

Traces:
* We still copy traces, we only do so once per table/conditional.
* Traces are created completely empty (a valid suffix trace of size 0).
* Traces are appended everytime a control is evaluated in reverse order (valid suffixes throughout by induction).
* MergeTracesOnCondition merges matches from one trace that do not have a corresponding match in the other trace.

Evalluation:
* Make very explicit how the guards are constructed for each entry, both in code and docs.
* Build the if-then-else symbolic expressions for table_index correctly inside-out in reverse order.
* Evaluating an action of an entry is guarded by the negation of all higher priority entries conditions, and the initial table guard.

konne88 reviewed (changes requested):
Overall, this look great!

On a different note. This CL is way too big :-) Let's not worry about it for this CL, just something to keep in mind. Each CL should be smaller (but each CL still has to build, and make a meaningful improvement to the code). From the commit history, it looks like you kept adding features to this CL. That's pretty much exactly the opposite of what you want to do :) If you want to add features on top of a CL that's already outstanding, create a new CL, don't modify the old one.

smolkaj commented:
In @KinanBab's defense regarding large "CLs", it turned out early in the internship while @konne88 was gone that creating chains of CLs and then amending them is challenging with git. We decided then that it's more important for @KinanBab to be reasonably productive than for the CLs to be as small as we would usually like them to be internally. We don't have the tool support externally that we're used to internally, so the tradeoffs are different.

KinanBab committed:
Address comments

* Rename GuardedMap to SymbolicGuardedMap
* Add the SymbolicPerPacketState type alias again and use it for internal evaluation functions
* Change SymbolicGuardedMap to use a factory
* Update certain outdated comments
* Use const refs for the ActionContext parameter in expression evaluation functions.
* Remove ActionContext and SYmbolicPerPacketState parameters for expression evaluation functions that do not need them.
* Change size_t .Count() to bool .ContainsKey() in SymbolicGuardedMap and update its usage
* Remove $ special value from headers, and instead evaluate mark_to_drop in a semantically corrrect way by setting egress_spec to x#1FF. Assert domain restriction on egress_spec.

KinanBab commented:
I believe I addressed all your comments.

Thanks to @smolkaj for coming to my defense. I still admit that I could have done a bit better on this front even with that limitation around chains of CLs.

KinanBab committed:
More clean up.

* Remove value from concrete and symbolic table matches.
* Remove SymbolicHeaders and use SymbolicPerPacketState instead.
* Rename ConcreteHeaders to ConcretePerPacketState for symmetry!

KinanBab committed:
Remove completed TODO
